File: go/day12part2.go
package main
/*
# PROBLEM:
#     Given a list of input instructions of form <I><Distance/Degrees>, where
#        I = N (North)| S (South) | E (East) | W (West)
#        I = L (Left) | R (Right) | F (Forward)
#
#     WHERE ALL INSTRUCTIONS REFER TO THE RELATIVE POSITION OF A WAYPOINT,
#     (N/E/S/W move the waypoint those positions, L/R rotate the waypoint about
#     the ship position)
#     with the exception of F which moves the ship to the waypoint <n> times
#    (the waypoint maintains it's relative position to the ship at all times)
#     And a starting position of (0,0), facing East, waypoint at (10 east, 1 north):
#        a) Calculate the final position by following the instructions
#        b) Compute the "Manhattan Distance" (sum of absolute east/west + north/south positions)
#
#   NOTE: Rotating a vector needs Maths. If you're doing it properly. But we can
#         cheat because all the angles given are multiples of 90 degrees. Fortunately.

This is going to be a reasonably straight translation from the Python version
of my answer too.
*/
import(
  "os"
  "log"
  "bufio"
  "fmt"
  "strings"
  "strconv"
)

//There's a limited set of instructions we support so enumerate them:
type Instruction int
const (
  North Instruction = iota
  East
  South
  West
  Left
  Right
  Forward
)

//We have 2 types of data, arguably both are just simple Maps but...
type ACommand struct {
  ins Instruction
  parm int
  valid bool
}
type StateVector struct {
  shipNorth int
  shipEast  int
  wpNorth   int
  wpEast    int
}

//-----------------------------------------------------------------------------
//Helper function for showing the Instructions:
func (i Instruction) String() string {
  return [...]string{"North","East","South","West","Left","Right","Forward"}[i]
}

// Oh wow, there's no built-in Abs function for ints...
func intAbs(x int) int {
  if x<0 {
    return -1 * x
  }
  return x
}


/*
-------------------------------------------------------------------------------
*/
func parseInstruction(insTxt string) ACommand {
  p,_ := strconv.Atoi(insTxt[1:])  //Extract the number parameter
  outCommand := ACommand{
    ins   : Forward,
    parm  : p,
    valid : true,
  }
  i := insTxt[0]
  switch i {
    case 'N' : { outCommand.ins = North }
    case 'E' : { outCommand.ins = East }
    case 'S' : { outCommand.ins = South }
    case 'W' : { outCommand.ins = West }
    case 'L' : { outCommand.ins = Left}
    case 'R' : { outCommand.ins = Right}
    case 'F' : { outCommand.ins = Forward}
    default  : {
      outCommand.valid = false;
      log.Fatal("Unrecognised Command: " + string(i))
    }
  }
  return outCommand
}


/*
-------------------------------------------------------------------------------
*/
func rotateWP(inState StateVector,dir Instruction,degrees int) StateVector {
  outState := inState
  //Some simplification:
  if degrees==270 {
    degrees=90
    if dir==Left {
      dir=Right
    } else {
      dir=Left
    }
  }
  //Now process the state vector update:
  if degrees==180 {
    outState.wpNorth = -1 * inState.wpNorth
    outState.wpEast  = -1 * inState.wpEast
  } else if degrees==90 {
    if dir==Left {
      outState.wpNorth = inState.wpEast
      outState.wpEast  = -1 * inState.wpNorth
    } else {
      outState.wpNorth = -1 * inState.wpEast
      outState.wpEast  = inState.wpNorth
    }
  } else {
    log.Fatal("Unrecognised Rotation Command")
  }

  return outState
}



/*
-------------------------------------------------------------------------------
*/
func updateStateVector(inState StateVector, cmd ACommand) StateVector {
  outState := inState
  switch cmd.ins {
    //These commands are "easy" and just move the waypoint:
    case North : { outState.wpNorth = inState.wpNorth + cmd.parm}
    case East  : { outState.wpEast  = inState.wpEast  + cmd.parm}
    case South : { outState.wpNorth = inState.wpNorth - cmd.parm}
    case West  : { outState.wpEast  = inState.wpEast  - cmd.parm}
    //Left and Right require transformation of the Waypoint vector
    case Left  : { outState = rotateWP(inState,cmd.ins,cmd.parm)}
    case Right  : { outState = rotateWP(inState,cmd.ins,cmd.parm)}
    //Forward is the only command that moves the ship, to the Waypoint <parm>
    //times:
    case Forward : {
      outState.shipNorth = inState.shipNorth + (cmd.parm * inState.wpNorth)
      outState.shipEast  = inState.shipEast  + (cmd.parm * inState.wpEast)

    }
    default : { log.Fatal("Cannot execute unsupported command")}
  }


  return outState
}


/*
-------------------------------------------------------------------------------
-------------------------------------------------------------------------------
-------------------------------------------------------------------------------
The instructions here are simple enough to not really justify separating
out the load-process-output phases, so main() carries more weight than normal.
*/
func main() {
  infile := "../data/day12_input.txt"
  //infile := "../data/day12_test.txt"

  //Here's the master state vector we'll update throughout.
  stateVector := StateVector {
    shipNorth : 0,
    shipEast  : 0,
    wpNorth   : 1,
    wpEast    : 10,
  }

  //boilerplate file handling:
  fh,err := os.Open(infile)
  if err != nil {
    log.Fatal(err)
  }
  defer fh.Close()
  s := bufio.NewScanner(fh)

  //Here's the main loop. Each input line is an instruction to follow so
  //we update the stateVector as we go

  validInstructions := 0

  for s.Scan() {
    insText := strings.TrimSuffix(s.Text(),"\n")
    if len(insText)>0 {
      command := parseInstruction(insText)
      fmt.Printf("(%5s):",insText)
      if command.valid {
        stateVector = updateStateVector(stateVector,command)
        fmt.Printf("%v\n", stateVector)
        validInstructions++
      }
    }
  }
  manhattanDistance := intAbs(stateVector.shipNorth) + intAbs(stateVector.shipEast)
  fmt.Printf("Final Ship Position after %d instructions (%04d,%04d) has Manhattan Distance = %d\n",
      validInstructions,
      stateVector.shipNorth,
      stateVector.shipEast,
      manhattanDistance,
  )
}

File: python/day12part2.py
#!/usr/bin/python3
#
#  My answer for Day 12 Part 2
#
# PROBLEM:
#     Given a list of input instructions of form <I><Distance/Degrees>, where
#        I = N (North)| S (South) | E (East) | W (West)
#        I = L (Left) | R (Right) | F (Forward)
#
#     WHERE ALL INSTRUCTIONS REFER TO THE RELATIVE POSITION OF A WAYPOINT,
#     (N/E/S/W move the waypoint those positions, L/R rotate the waypoint about
#     the ship position)
#     with the exception of F which moves the ship to the waypoint <n> times
#    (the waypoint maintains it's relative position to the ship at all times)
#     And a starting position of (0,0), facing East, waypoint at (10 east, 1 north):
#        a) Calculate the final position by following the instructions
#        b) Compute the "Manhattan Distance" (sum of absolute east/west + north/south positions)
#
#   NOTE: Rotating a vector needs Maths. If you're doing it properly. But we can
#         cheat because all the angles given are multiples of 90 degrees. Fortunately.
import sys
import copy

################################################################################
def parse_instruction(txt) :
    retIns = { 'ins' : "X", 'parm': 0}
    txt = txt.strip()
    if len(txt) > 0 :
        retIns['ins'] = txt[0]
        retIns['parm'] = int(txt[1:])

    return retIns
################################################################################
def rotateWaypoint(inState,dir,degrees) :
    #This would be tough and require some real Trig if the input contained
    #anything other than multiples of 90 degrees. But it doesn't, so we can
    #cheat :-)
    outState = copy.copy(inState)

    #First, simplify the direction:
    if degrees==270 :
        degrees=90
        if dir=="L" :
            dir="R"
        elif dir=="R" :
            dir="L"

    #Now implement the transformation on the waypoint:
    if degrees==180 :
        outState['wpNorth'] = -1 * inState['wpNorth']
        outState['wpEast']  = -1 * inState['wpEast']
    elif degrees==90 and dir=="L" :
        outState['wpNorth'] = inState['wpEast']
        outState['wpEast']  = -1 * inState['wpNorth']
    elif degrees==90 and dir=="R" :
        outState['wpNorth'] = -1 * inState['wpEast']
        outState['wpEast']  = inState['wpNorth']
    else :
        print(f"ARRRRGH unsupported rotation {dir}{degrees}")

    return outState

################################################################################
def update_state_vector(inState,ins) :
    outState = copy.copy(inState)
    if ins['ins'] == 'N' :
        outState['wpNorth'] = inState['wpNorth'] + ins['parm']
    elif ins['ins'] == 'E' :
        outState['wpEast'] = outState['wpEast'] + ins['parm']
    elif ins['ins'] == 'S' :
        outState['wpNorth'] = outState['wpNorth'] - ins['parm']
    elif ins['ins'] == 'W' :
        outState['wpEast'] = outState['wpEast'] - ins['parm']
    elif ins['ins'] == 'L' or ins['ins'] == 'R' :
        #These are the complex instructions in Part 2. Requires translating
        #the Waypoint around the ship by L/R <degrees>
        outState = rotateWaypoint(inState,ins['ins'],ins['parm'])
    elif ins['ins'] == 'F' :
        #This is the only time the ship moves in Part 2. Move by
        #offset waypoint coordinates X times:
        outState['shipNorth'] = inState['shipNorth'] + (ins['parm'] * inState['wpNorth'])
        outState['shipEast']  = inState['shipEast'] + (ins['parm'] * inState['wpEast'])

    else :
        print(f"Argh, unsupported instruction {ins['ins']}")

    return outState

################################################################################
################################################################################
################################################################################
if __name__ == "__main__":
    stateVector = { 'shipNorth': 0, 'shipEast': 0, 'wpNorth': 1, 'wpEast': 10 }

    infile = "../data/day12_input.txt"
    #infile = "../data/day12_test.txt"

    validInstructions = 0
    with open(infile,"r") as insfl :
        for ins in insfl :
            instruction=parse_instruction(ins)
            if instruction['ins'] == "X" : continue #duff instruction
            validInstructions = validInstructions + 1
            print(f"({instruction['ins']}{instruction['parm']:4})",end='->')
            stateVector = update_state_vector(stateVector,instruction)
            print(stateVector)

    print(f"Final State Vector after {validInstructions} instructions",end=': ')
    print(stateVector)
    print("Manhattan Distance: {}".format(abs(stateVector['shipNorth'])+abs(stateVector['shipEast'])))
